### PR TITLE
[core] Fix proposals shuffling implementation

### PR DESCRIPTION
**Pull request by robin-near**
The SliceRandom::shuffle implementation changed from rand 0.6.5 to 0.7.0 in a subtle way: https://github.com/rust-random/rand/pull/809 because the underlying gen_index implementation changed. (It appears that @matklad 's concern on #7717 turns out to be quite real!)

In order to reproduce the old behavior and to make sure this never changes again, this PR replicates the gen_index implementation (it's pretty straight-forward) from 0.6.5. The test case is updated (the old test case was unfortunately incorrect from #7717). Also, @mm-near 's #7909 now passes with this change.

### CODE REVIEW

File: chain/epoch-manager/src/proposals.rs [hunk 1]
```diff
@@ -270,15 +269,51 @@ mod old_validator_selection {
...
+        let zone = u64::MAX - (u64::MAX - bound + 1) % (bound);
+        loop {
+            let v = rng.next_u64();
+            let mul = (v as u128) * (bound as u128);
+            let (hi, lo) = ((mul >> 64) as u64, mul as u64);
+            if lo < zone {
+                return hi;
+            }
+        }
+    }
```
Comment (matklad) [inline chain/epoch-manager/src/proposals.rs:289]: If we are feeling enterpisey, I'd probably extracted this code into a separate module/type `EpochManagerRng` to make sure its public API *doesn't* expose `rand`. Right now this code has an implicit invariant of "don't use any rand methods besides from_seed and next_u64", and I'd love for that invariant to have teeth.

File: chain/epoch-manager/src/proposals.rs [hunk 2]
```diff
@@ -270,15 +269,51 @@ mod old_validator_selection {
...
+
+        #[test]
+        pub fn proposal_shuffling_sanity_check_one_element() {
+            let mut dup_proposals = vec![8];
+            shuffle_duplicate_proposals(
+                &mut dup_proposals,
+                CryptoHash::hash_bytes(&[1, 2, 3, 4, 5]).as_bytes().clone(),
+            );
+            assert_eq!(dup_proposals, vec![8]);
+        }
```
Comment (matklad) [inline chain/epoch-manager/src/proposals.rs:315]: I'd write his as a randomized property test: 

```
for len in 0..10 {
  let dup_proposals = random_vec_of_len_len
}
```

**Review by matklad — APPROVED** [review]

**Review by mm-near — APPROVED** [review]

File: chain/epoch-manager/src/proposals.rs (hunk 1 above)
Comment (robin-near) [reply chain/epoch-manager/src/proposals.rs]: Sorry not exactly sure what you mean - the module is already named old_* and these two functions are not public.f

Comment (matklad) [reply chain/epoch-manager/src/proposals.rs]: Ah, sorry, I missed two points:

* that this is already a separate module
* that we no longer relying on randomness here for the latest protocol

I was fearing that we still use `rand`, and need to guard aginst future code changes.

**Pull request merged**